File: 540. Single Element in a Sorted Array.py
class Solution:
    def singleNonDuplicate(self, nums: List[int]) -> int:
        lo, hi = 0, len(nums)-1
        while lo <= hi:
            if lo == hi:
                return nums[lo]
            mid = (hi-lo)//2+lo
            if mid%2 == 0:
                if mid+1 == len(nums):
                    return nums[mid]
                else:
                    if nums[mid] == nums[mid+1]:
                        lo = mid+2
                    else:
                        hi = mid
            else:
                if nums[mid] == nums[mid-1]:
                    lo = mid+1
                else:
                    hi = mid-1

File: 540. single element in a sorted array.java
class Solution {
    public int singleNonDuplicate(int[] nums) {
        if (nums==null || nums.length==0) {
            return 0;
        }
        int lo=0, hi=nums.length-1;
        while (lo <= hi) {
            int mid = (hi-lo)/2+lo;
            if (mid<nums.length-1 && mid%2==0 && nums[mid]==nums[mid+1]) {
                // the single one comes after this
                lo = mid+1;
            } else if (mid<nums.length-1 && mid%2!=0 && nums[mid]==nums[mid+1]) {
                // the single one comes before this
                hi = mid-1;
            } else if (mid>0 && mid%2==0 && nums[mid]==nums[mid-1]) {
                // the single one comes before this
                hi = mid-1;
            } else if (mid>0 && mid%2!=0 && nums[mid]==nums[mid-1]) {
                // the single one comes after this
                lo = mid+1;
            } else {
                return nums[mid];
            }
        }
        return 0;
    }
}